### PR TITLE
util: fix cf disorder when creating cfs

### PR DESCRIPTION
**Pull request by hhkbp2**
Hi,

This PR tries to fix the cf disorder problem, which will cause failure on applying sst format snapshot, like these logs indicate:

```
2017/04/26 12:44:01.436 region.rs:241: [ERROR] failed to apply snap: Other(StringError(“Invalid argument: External file column family id dont match”))!!!
2017/04/26 12:44:01.436 region.rs:175: [INFO] [region 885] begin apply snap data
2017/04/26 12:44:01.783 region.rs:241: [ERROR] failed to apply snap: Other(StringError(“Invalid argument: External file column family id dont match”))!!!
```

PTAL @zhangjinpeng1987 @siddontang @BusyJay

### CODE REVIEW

File: src/util/mod.rs [hunk 1]
```diff
@@ -467,6 +467,11 @@ impl<T> RingQueue<T> {
     }
 }
 
+// `cfs_diff' Returns a Vec of cf which is in `a' but not in `b'.
+pub fn cfs_diff<'a, 'b>(a: &'a Vec<&'b str>, b: &'a Vec<&'b str>) -> Vec<&'b str> {
```
Comment (siddontang) [inline src/util/mod.rs]: please add a test to cover it individually.

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -692,11 +692,12 @@ fn build_raftkv(config: &toml::Value,
     let trans = ServerTransport::new(ch);
     let path = Path::new(&cfg.storage.path).to_path_buf();
     let db_opts = get_rocksdb_db_option(config);
-    let mut cfs_opts = HashMap::default();
-    cfs_opts.insert(CF_DEFAULT, get_rocksdb_default_cf_option(config, total_mem));
-    cfs_opts.insert(CF_LOCK, get_rocksdb_lock_cf_option(config));
-    cfs_opts.insert(CF_WRITE, get_rocksdb_write_cf_option(config, total_mem));
-    cfs_opts.insert(CF_RAFT, get_rocksdb_raftlog_cf_option(config, total_mem));
+    let cfs_opts = vec![
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: is it better to sort the vec with name?

Comment (hhkbp2) [inline src/bin/tikv-server.rs]: It's better to keep the order as it's for compatibility.

File: src/util/mod.rs (hunk 1 above)
Comment (hhkbp2) [inline src/util/mod.rs]: address comment.

**Comment by hhkbp2** [comment]
Please notice: until we know for sure that all the existing RocksDB instances have the same cf order, sst file format snapshot could not be enabled for a version upgrade.
@zhangjinpeng1987 @siddontang @BusyJay

File: src/util/mod.rs [hunk 3]
```diff
@@ -467,6 +467,11 @@ impl<T> RingQueue<T> {
     }
 }
 
+// `cfs_diff' Returns a Vec of cf which is in `a' but not in `b'.
+pub fn cfs_diff<'a, 'b>(a: &'a [&'b str], b: &'a [&'b str]) -> Vec<&'b str> {
```
Comment (siddontang) [inline src/util/mod.rs]: do we need lifetime `'a` here?

File: src/util/rocksdb.rs [hunk 4]
```diff
@@ -84,16 +95,16 @@ fn check_and_open(path: &str,
 
     // List all column families in current db.
     let cfs_list = try!(DB::list_column_families(&db_opt, path));
-    let existed: HashSet<&str> = cfs_list.iter().map(|v| v.as_str()).collect();
-    let needed: HashSet<&str> = cfs_opts.keys().map(|v| *v).collect();
+    let existed: Vec<&str> = cfs_list.iter().map(|v| v.as_str()).collect();
+    let needed: Vec<&str> = cfs_opts.iter().map(|cf_options| cf_options.cf).collect();
```
Comment (siddontang) [inline src/util/rocksdb.rs]: g/cf_options/cf_opt/s

File: src/util/rocksdb.rs [hunk 5]
```diff
@@ -45,36 +45,47 @@ pub fn open_opt(opts: Options,
...
+        if let Some(cf_options) = cfs_opts.iter().find(|x| x.cf == CF_DEFAULT) {
             cfs.push(CF_DEFAULT);
-            cfs_opts_ref.push(opt);
+            cfs_opts_ref.push(&cf_options.options);
         }
         let mut db = try!(DB::open_cf(db_opt, path, cfs.as_slice(), cfs_opts_ref.as_slice()));
-        for (cf, cf_opt) in &cfs_opts {
-            if *cf == "default" {
+        for cf_options in &cfs_opts {
+            if cf_options.cf == "default" {
```
Comment (siddontang) [inline src/util/rocksdb.rs]: we can use CF_DEFAULT here

File: src/util/rocksdb.rs (hunk 4 above)
Comment (siddontang) [inline src/util/rocksdb.rs]: I am confused with cfs_opts and cf_options now. maybe we should think a better name for cf_options, I even prefer `v` here directly.

**Comment by zhangjinpeng1987** [comment]
To compatible with old version, I think we should set cf=null when create SstFilterWriter to ignore this error.

**Comment by siddontang** [comment]
I guess maybe we can let RocksDB only check name but ignore the ID.

File: src/bin/tikv-server.rs (hunk 2 above)
Comment (siddontang) [inline src/bin/tikv-server.rs]: do we need vector now?

Comment (hhkbp2) [inline src/bin/tikv-server.rs]: IMO it's better to use `Vec` to preserve the order even though the cf order would not be a problem to sst format snapshot applying now. It's usually good choice to use a simple `Vec` when it's not performance sensitive.

**Comment by hhkbp2** [comment]
PTAL @zhangjinpeng1987 @siddontang

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

**Comment by siddontang** [comment]
Please fix conflict @hhkbp2

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/util/rocksdb.rs [hunk 6]
```diff
@@ -205,31 +214,31 @@ impl SliceTransform for FixedPrefixSliceTransform {
...
+    use super::{check_and_open, CFOptions};
 
     #[test]
     fn test_check_and_open() {
         let path = TempDir::new("_util_rocksdb_test_check_column_families").expect("");
         let path_str = path.path().to_str().unwrap();
 
         // create db when db not exist
-        let cfs_opts = map!["default" => Options::new()];
+        let cfs_opts = vec![CFOptions::new("default", Options::new())];
```
Comment (ngaut) [inline src/util/rocksdb.rs]: Can we define the "default" string to constant.

Comment (hhkbp2) [inline src/util/rocksdb.rs]: There is a var `CF_DEFAULT` defined to be "default", which is used everywhere. We could use that.

**Comment by hhkbp2** [comment]
PTAL @ngaut

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
CI failed @hhkbp2

**Comment by hhkbp2** [comment]
@siddontang I don't have a clear idea how to interact with travis-ci to debug this issue. Any help?

**Pull request merged**